src/lib/navigation/navigation-mode-factory.provider.ts

```diff
@@ -0,0 +1,78 @@
+import {FreeNavigationMode} from './free-navigation-mode';
+import {NavigationMode} from './navigation-mode.interface';
+import {SemiStrictNavigationMode} from './semi-strict-navigation-mode';
+import {StrictNavigationMode} from './strict-navigation-mode';
+import {WizardComponent} from '../components/wizard.component';
+import {NavigationModeInput} from './navigation-mode-input.interface';
+
+/**
+ * A factory used to create [[NavigationMode]] instances
+ */
+export class BaseNavigationModeFactory {
+
+  /**
+   * @inheritDoc
+   */
+  create(wizard: WizardComponent, navigationModeInput: NavigationModeInput): NavigationMode {
+    let navigationModeName: string;
+    if (typeof navigationModeInput === 'function') {
+      // input is a function
+      const navigationMode = navigationModeInput(wizard);
+      if (typeof navigationMode === 'string') {
+        // function returned a name
+        navigationModeName = navigationMode;
```

Comment:
@madoar , I wonder if we really need an option to return a string from the function. Perhaps, the user could just construct the necessary Strict|SemiStrict|FreeNavigationMode?

Removing this signature will simplify this code and also the type declaration of NavigationModeInput:

```diff
-export type NavigationModeInput = string|((wizard: WizardComponent) => string|NavigationMode);
+export type NavigationModeInput = string|((wizard: WizardComponent) => NavigationMode);
```

Reply:
And of course it will make things more understandable for library users.

Reply:
I think it is fine to change the method like you described to return a `NavigationMode` instance.
I don't think that we limit the library users in any way if we do this.

Reply:
Done in feecb85